pkg/node/node.go

```diff
@@ -101,105 +103,261 @@ func createAddressFilter(
+		// retry here.
+		return n.publishSignature(ctx, keepAddress, digest, signature)
+	}
+}
 
-	if err := n.ethereumChain.SubmitSignature(keepAddress, signature); err != nil {
+// publisSignature takes the provided signature and attempts to publish it to
+// the chain. It implements retry mechanism allowing to attempt to publish again
+// in case of a failure.
+//
+// We do implement a retry in this function because the retry mechanism is much
+// more complex than in case of e.g. publishSignerPublicKey. Although all keep
+// members are supposed to try to publish the signature, only one transaction
+// succeeds. For each attempt, we need to check if the keep still awaits
+// a signature. Also, we need to implement some sane delay between attempts so
+// that we do not waste gas.
+func (n *Node) publishSignature(
+	ctx context.Context,
+	keepAddress common.Address,
+	digest [32]byte,
+	signature *ecdsa.Signature,
+) error {
+	attemptCounter := 0
+	for {
+		attemptCounter++
+
+		// Check if keep still awaits a signature for this digest.
+		// We do this check here in case the attempt was retried because of
+		// on-chain failure during submission. In this case we want to make sure
+		// no other member published the signature in the meantime so that we
+		// do not burn ether on redundant submission.
+		//
+		// If the check failed, we retry from the beginning.
 		isAwaitingSignature, err := n.ethereumChain.IsAwaitingSignature(keepAddress, digest)
 		if err != nil {
-			return fmt.Errorf("failed to verify if keep is still awaiting signature: [%v]", err)
+			logger.Errorf(
+				"failed to verify if keep [%s] is still awaiting signature: [%v]",
+				keepAddress.String(),
+				err,
+			)
+			continue
 		}
 
+		// Someone submitted the signature and it was accepted by the keep.
+		// We are fine, leaving.
 		if !isAwaitingSignature {
-			logger.Infof("signature submitted by another member: [%+x]", digest)
-
+			logger.Infof(
+				"signature for keep [%s] already submitted: [%+x]",
+				keepAddress.String(),
+				digest,
+			)
 			return nil
 		}
 
-		return fmt.Errorf("failed to submit signature: [%v]", err)
-	}
+		logger.Infof(
+			"publishing signature for keep [%s]; attempt [%v]",
+			keepAddress.String(),
+			attemptCounter,
+		)
 
-	logger.Infof("submitted signature for digest: [%+x]", digest)
+		if err := n.ethereumChain.SubmitSignature(keepAddress, signature); err != nil {
+			isAwaitingSignature, err := n.ethereumChain.IsAwaitingSignature(keepAddress, digest)
+			if err != nil {
+				logger.Errorf("failed to verify if keep is still awaiting signature: [%v]", err)
+				continue
+			}
+
+			// Check if we failed because someone else submitted in the meantime
+			// or because something wrong happened with our transaction.
+			if !isAwaitingSignature {
+				logger.Infof(
+					"signature for keep [%s] already submitted: [%+x]",
+					keepAddress.String(),
+					digest,
+				)
+				return nil
+			}
```

Comment:
Maybe we don't need this part as we're going to retry from the beginning and we've got this check starting in L303?

Reply:
I was thinking about it and I also don't like the code duplication here but I think this part is what's going to be executed in the majority of cases and it's nice to do not wait one minute with logging it. If we have a one-second delay, I'd remove it but for one minute, I think it makes sense to have it here.